src/main/java/mate.academy/lib/Injector.java

```diff
@@ -1,13 +1,81 @@
 package mate.academy.lib;
 
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.util.HashMap;
+import java.util.Map;
+import mate.academy.service.FileReaderService;
+import mate.academy.service.ProductParser;
+import mate.academy.service.ProductService;
+import mate.academy.service.impl.FileReaderServiceImpl;
+import mate.academy.service.impl.ProductParserImpl;
+import mate.academy.service.impl.ProductServiceImpl;
+
 public class Injector {
     private static final Injector injector = new Injector();
+    private final Map<Class<?>, Object> instances = new HashMap<>();
 
     public static Injector getInjector() {
         return injector;
     }
 
     public Object getInstance(Class<?> interfaceClazz) {
-        return null;
+        if (interfaceClazz.isAnnotationPresent(Component.class)) {
+            Object clazzImplementationInstance = null;
+            Class<?> clazz = findImplementation(interfaceClazz);
+            Field[] declaredFields = clazz.getDeclaredFields();
+            for (Field field : declaredFields) {
+                if (field.isAnnotationPresent(Inject.class)) {
+                    Object fieldInstance = getInstance(field.getType());
+                    clazzImplementationInstance = createNewInstance(clazz);
+                    field.setAccessible(true);
+                    try {
+                        field.set(clazzImplementationInstance, fieldInstance);
+                    } catch (IllegalAccessException e) {
+                        throw new RuntimeException("Can't initialize field value. Class: "
+                                + clazz.getName() + ". Field: " + field.getName(), e);
+                    }
+                }
+            }
+            if (clazzImplementationInstance == null) {
+                clazzImplementationInstance = createNewInstance(clazz);
+
+            }
+            return clazzImplementationInstance;
+
+        } else {
+            throw new RuntimeException("This instance cannot be created, because class has"
+                    + "no annotation '@Component'");
+        }
+    }
```

Comment:
This method does not check the instances map before creating a new instance as suggested by the checklist. To optimize, you should check if the instance already exists in the instances map before creating a new one.

Reply:
Actually, the checking is in the 'createNewInstance' method. Why it can't be there ?